pkg/component/runtime/service.go

```diff
@@ -0,0 +1,441 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package runtime
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent/pkg/component"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+	"github.com/kardianos/service"
+)
+
+type serviceAction int
+
+const (
+	serviceStart serviceAction = iota
+	serviceStop
+	serviceTeardown
+)
+
+const (
+	checkServiceStatusInterval = 30 * time.Second // 30 seconds default for now, consistent with the command check-in interval
+)
+
+var (
+	ErrOperationSpecUndefined = errors.New("operation spec undefined")
+	ErrInvalidServiceSpec     = errors.New("invalid service spec")
+)
+
+// ServiceRuntime provides the command runtime for running a component as a service.
+type ServiceRuntime struct {
+	current component.Component
+	log     *logger.Logger
+
+	ch       chan ComponentState
+	actionCh chan serviceAction
+	compCh   chan component.Component
+	statusCh chan service.Status
+
+	state          ComponentState
+	lastCheckin    time.Time
+	missedCheckins int
+
+	lastServiceStatus service.Status
+}
+
+// NewServiceRuntime creates a new command runtime for the provided component.
+func NewServiceRuntime(comp component.Component, logger *logger.Logger) (ComponentRuntime, error) {
+	if comp.Spec.Spec.Service == nil {
+		return nil, errors.New("must have service defined in specification")
+	}
+
+	return &ServiceRuntime{
+		current:           comp,
+		log:               logger,
+		ch:                make(chan ComponentState),
+		actionCh:          make(chan serviceAction),
+		compCh:            make(chan component.Component),
+		statusCh:          make(chan service.Status),
+		state:             newComponentState(&comp),
+		lastServiceStatus: service.StatusUnknown,
+	}, nil
+}
+
+// check executes the service check command
+func (s *ServiceRuntime) check(ctx context.Context) error {
+	if s.current.Spec.Spec.Service.Operations.Check == nil {
+		s.log.Errorf("missing check spec for %s service", s.current.Spec.BinaryName)
+		return ErrOperationSpecUndefined
+	}
+	s.log.Debugf("check if the %s is installed", s.current.Spec.BinaryName)
+	return executeServiceCommand(ctx, s.current.Spec.BinaryPath, s.current.Spec.Spec.Service.Operations.Check)
+}
+
+// install executes the service install command
+func (s *ServiceRuntime) install(ctx context.Context) error {
+	if s.current.Spec.Spec.Service.Operations.Install == nil {
+		s.log.Errorf("missing install spec for %s service", s.current.Spec.BinaryName)
+		return ErrOperationSpecUndefined
```

Comment:
Same as above.

Reply:
I'll leave these changes in place for install and uninstall